File: src/main/java/dev/matheus/task/configs/infra/Swagger.java
package dev.matheus.task.configs.infra;

import io.swagger.v3.oas.models.Components;
import io.swagger.v3.oas.models.OpenAPI;
import io.swagger.v3.oas.models.info.Contact;
import io.swagger.v3.oas.models.info.Info;
import io.swagger.v3.oas.models.info.License;
import io.swagger.v3.oas.models.security.SecurityRequirement;
import io.swagger.v3.oas.models.security.SecurityScheme;
import org.springdoc.core.models.GroupedOpenApi;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

@Configuration
public class Swagger {

    @Bean
    public GroupedOpenApi publicApi(){
        return GroupedOpenApi.builder()
                .group("task")
                .displayName("Tasks API V1")
                .pathsToMatch("/**")
                .build();
    }

    private SecurityScheme createAPIKeyScheme() {
        return new SecurityScheme().type(SecurityScheme.Type.HTTP)
                .bearerFormat("JWT")
                .scheme("bearer");
    }

    @Bean
    public OpenAPI openAPI() {
        return new OpenAPI().addSecurityItem(new SecurityRequirement().
                        addList("Bearer Authentication"))
                .components(new Components().addSecuritySchemes
                        ("Bearer Authentication", createAPIKeyScheme()))
                .info(new Info()
                        .title("Desafio Back-end Código-Certo")
                        .description("API REST para gerenciamento de tarefas, desenvolvido para o processo seletivo da Código-Certo, onde o usuário deve se registrar e autenticar para realizar o CRUD de tarefas.")
                        .contact(new Contact()
                                .name("@matheus")
                                .email("matheusgmello1@gmail.com"))
                        .license(new License()
                                .name("MIT LICENSE")
                                .url("https://github.com/matheusgmello")));
    }
}

File: swagger.ts
import swaggerAutogen from "swagger-autogen";

const doc = {
  info: {
    version: "v1.0.0",
    title: "Task API - Código Certo",
    description:
      "API para gestão de tarefas com o recurso de autenticação de usuário",
  },
  servers: [
    {
      url: "http://localhost:8080",
      description: "",
    },
  ],
  components: {
    securitySchemes: {
      bearerAuth: {
        type: "http",
        scheme: "bearer",
      },
    },
  },
};

const outputFile = "./swagger_output.json";
const endpointsFiles = [
  "./src/routes/task.routes.ts",
  "./src/routes/user.routes.ts",
];

swaggerAutogen({ openapi: "3.0.0" })(outputFile, endpointsFiles, doc);
